File: backend/sangam_mw/api/routes/lineage.py
"""
Lineage & impact analysis API.

GET /api/v1/lineage/{flow_id}        flow's full lineage graph
GET /api/v1/lineage/impact           affected flows for a given resource
"""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from ...db.base import get_session
from ...lineage.tracker import get_flow_lineage, impact_analysis
from ...rbac.permissions import Permission, require_permission

router = APIRouter(prefix="/lineage", tags=["lineage"])


@router.get("/{flow_id}")
async def flow_lineage(
    flow_id: str,
    _user: dict = Depends(require_permission(Permission.FLOW_READ)),
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    edges = await get_flow_lineage(session, flow_id)
    sources = [e for e in edges if e["direction"] == "read"]
    sinks = [e for e in edges if e["direction"] == "write"]
    return {
        "flow_id": flow_id,
        "sources": sources,
        "sinks": sinks,
        "total_edges": len(edges),
    }


@router.get("/impact")
async def impact(
    connection_id: str | None = Query(None),
    connector_id: str | None = Query(None),
    object_name: str | None = Query(None),
    _user: dict = Depends(require_permission(Permission.FLOW_READ)),
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    affected = await impact_analysis(
        session,
        connection_id=connection_id,
        connector_id=connector_id,
        object_name=object_name,
    )
    return {
        "query": {
            "connection_id": connection_id,
            "connector_id": connector_id,
            "object_name": object_name,
        },
        "affected_flows": affected,
        "count": len(affected),
    }

File: frontend/src/api/lineage.ts
import { api } from './client'

export interface LineageEdge {
  direction: 'read' | 'write'
  connection_id: string
  connector_id: string
  object_name: string
  step_id: string
  run_id?: string
}

export interface FlowLineage {
  flow_id: string
  sources: LineageEdge[]
  sinks: LineageEdge[]
  total_edges: number
}

export interface ImpactResult {
  query: Record<string, string | null>
  affected_flows: string[]
  count: number
}

export const lineageApi = {
  getFlowLineage: (flowId: string) =>
    api.get<FlowLineage>(`/lineage/${flowId}`),
  impactAnalysis: (params: {
    connection_id?: string
    connector_id?: string
    object_name?: string
  }) => {
    const qs = new URLSearchParams(
      Object.fromEntries(
        Object.entries(params).filter(([, v]) => v != null) as [string, string][]
      )
    ).toString()
    return api.get<ImpactResult>(`/lineage/impact${qs ? `?${qs}` : ''}`)
  },
}
